lib/web/automaticupgrades.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package web
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/mod/semver"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/constants"
+	versionlib "github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/version"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+)
+
+const defaultChannelTimeout = 5 * time.Second
+
+// automaticUpgrades implements a version server in the Teleport Proxy.
+// It is configured through the Teleport Proxy configuration and tells agent updaters
+// which version they should install.
+func (h *Handler) automaticUpgrades(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	if h.cfg.AutomaticUpgradesChannels == nil {
+		return nil, trace.AccessDenied("This proxy is not configured to server automatic upgrades channels.")
+	}
+
+	// The request format is "<channel name>/{version,critical}"
+	// As <channel name> might contain "/" we have to split, pop the last part
+	// and re-construct the channel name.
+	channelAndType := p.ByName("request")
+
+	reqParts := strings.Split(strings.Trim(channelAndType, "/"), "/")
+	if len(reqParts) < 2 {
+		return nil, trace.BadParameter("path format should be /webapi/automaticupgrades/<channel>/{version,critical}")
```

Comment:
If you do decide to take my suggestion above and include /channel in the path, update this message as well.